src/Autofac/Core/Resolving/SegmentedStack.cs

```diff
@@ -0,0 +1,239 @@
+﻿// This software is part of the Autofac IoC container
+// Copyright © 2020 Autofac Contributors
+// https://autofac.org
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Autofac.Core.Resolving
+{
+    /// <summary>
+    /// Implements a segmented stack of items, which functions like a regular <see cref="Stack{T}"/>, but allows segments
+    /// of the stack to be enumerated without including items pushed before the segment.
+    /// </summary>
+    /// <typeparam name="T">The item type.</typeparam>
+    internal sealed class SegmentedStack<T> : IEnumerable<T>
+        where T : class
+    {
+        private T[] _array;
+        private int _next;
+        private int _activeSegmentBase;
+
+        private const int InitialCapacity = 16;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SegmentedStack{T}"/> class.
+        /// </summary>
+        public SegmentedStack()
+        {
+            _array = new T[InitialCapacity];
+        }
+
+        /// <summary>
+        /// Push an item onto the stack.
+        /// </summary>
+        /// <param name="item">The item.</param>
+        public void Push(T item)
+        {
+            // No null check for item here; internally called method only, known to never be null, and is a very hot path.
+            var next = _next;
```

Comment:
Is there a reason to assign local variables to the instance variables? (Like, is there some memory or perf optimization I'm unaware of? Not pointed - legit question.)

Reply:
I believe (but don't know all the detail tbh), that reading local variables on the stack can be faster than accessing heap variables multiple times, and can be worth it in these super-hot paths. Local variables can be cached more readily at the CPU, because the compiler understands the entire flow of the variable access (can't be changed by another thread or anything). Like I said, this may be off-base.

I have seen this in some of the internal collection methods in the dotnet runtime though, so it can't be too far off-base.

Fun fact, you'll see the odd uint casts in a couple of places, which allows the jitter to bypass array bounds checking I believe (other trick from the runtime).